AppLovin MAX Demo App - Swift/AppLovin MAX Demo App - Swift/Base Classes/MAAdViewSwiftUIWrapper.swift

```diff
@@ -0,0 +1,155 @@
+    {
+        let adView = MAAdView(adUnitIdentifier: adUnitIdentifier, adFormat: adFormat, sdk: sdk)
+        
+        adView.delegate = context.coordinator
+        adView.requestDelegate = context.coordinator
+        adView.revenueDelegate = context.coordinator
+        
+        adView.loadAd()
+        
+        return adView
+    }
+    
+    func updateUIView(_ uiView: MAAdView, context: Context) {}
+    
+    func makeCoordinator() -> Coordinator
+    {
+        Coordinator(parent: self)
+    }
+}
+
+@available(iOS 13.0, *)
+extension MAAdViewSwiftUIWrapper
+{
+    class Coordinator: NSObject, MAAdViewAdDelegate, MAAdRequestDelegate, MAAdRevenueDelegate
+    {
+        private let parent: MAAdViewSwiftUIWrapper
+        
+        init(parent: MAAdViewSwiftUIWrapper)
+        {
+            self.parent = parent
+        }
+        
+        func didStartAdRequest(forAdUnitIdentifier adUnitIdentifier: String)
+        {
+            parent.didStartAdRequest?(adUnitIdentifier)
+        }
+        
+        func didLoad(_ ad: MAAd)
+        {
+            parent.didLoad?(ad)
+        }
+        
+        func didFailToLoadAd(forAdUnitIdentifier adUnitIdentifier: String, withError error: MAError)
+        {
+            parent.didFailToLoadAd?(adUnitIdentifier, error)
+        }
+        
+        func didDisplay(_ ad: MAAd)
+        {
+            parent.didDisplay?(ad)
+        }
+        
+        func didFail(toDisplay ad: MAAd, withError error: MAError)
+        {
+            parent.didFailToDisplayAd?(ad, error)
+        }
+        
+        func didClick(_ ad: MAAd)
+        {
+            parent.didClick?(ad)
+        }
+        
+        func didExpand(_ ad: MAAd)
+        {
+            parent.didExpand?(ad)
+        }
+        
+        func didCollapse(_ ad: MAAd)
+        {
+            parent.didCollapse?(ad)
+        }
+        
+        func didHide(_ ad: MAAd)
+        {
+            parent.didHide?(ad)
+        }
+        
+        func didPayRevenue(for ad: MAAd)
+        {
+            parent.didPayRevenue?(ad)
+        }
+    }
+}
+
+@available(iOS 13.0, *)
+extension MAAdViewSwiftUIWrapper
+{
+    func deviceSpecificFrame() -> some View
+    {
+        modifier(MAAdViewFrame(adFormat: adFormat))
+    }
+}
+
+@available(iOS 13.0.0, *)
+struct MAAdViewFrame: ViewModifier
+{
+    let adFormat: MAAdFormat
+    private let isPhone = UIDevice.current.userInterfaceIdiom == .phone
+    
+    func body(content: Content) -> some View
+    {
+        if ( adFormat == .banner )
+        {
+            content
+                .frame(width: isPhone ? MAAdFormat.banner.size.width : MAAdFormat.leader.size.width,
+                       height: isPhone ? MAAdFormat.banner.size.height : MAAdFormat.leader.size.height)
+        }
+        else // adFormat == .mrec
+        {
+            content
+                .frame(width: 300, height: 250)
+        }
+    }
+}
+
+struct MAAdViewCallbackTableItem: Identifiable
```

Comment:
Is it not possible to maintain this callback array with just strings? I know SwiftUI requires elements to be Identifiable, but I don't remember if String qualifies for that.

If we need to keep this, I would make the name general: `CallbackTableItem`. We can continue to use this in other SwiftUI implementations, and this isn't specific to MAAdView.

Reply:
String doesn't - it would have to be extended as Identifiable or create a unique id each time from the string itself, which aren't appealing. I think it would have to be the latter.